### PR TITLE
Move eject drive and eject disk to new contextmenu system

### PR DESCRIPTION
**Pull request by razzeee**
## Description

Marvel your eyes, as your now able to eject your disks and drives right from your homescreen 🚀 

## Motivation and Context

No ejection from the homescreen ☮️ 

## How Has This Been Tested?

Only tested eject disk, as I have no drive to eject in my reach.

## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by ksooo** [comment]
Not sure whether it is a good idea to introduce the new cpp and header file in the xbmc source root folder.

**Comment by ksooo** [comment]
Otherwise, this is a nice one. 

But who nowadays is still using physical disks?. /me runs ;-)

**Comment by Razzeee** [comment]
I'm just moving it because it's there, I'm not using it either. 
And yes, if you have a better idea then a class in root, let me know. We will have more cases where context menus are not addon/video/music specific

**Comment by ksooo** [comment]
>  if you have a better idea then a class in root, let me know.

 sorry, no idea. maybe it is even okay to have it in the root directory for "global" context menus. @tamland ?

**Comment by Razzeee** [comment]
jenkins build this please

**Comment by Razzeee** [comment]
jenkins build this please

**Comment by Razzeee** [comment]
jenkins build this please

**Comment by ksooo** [comment]
@Razzeee hmm, why the hurry? Imo, this one needs to be reworked. Comments follow...

File: xbmc/ContextMenus.cpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+#include "guilib/GUIWindowManager.h"
+#include "storage/MediaManager.h"
+
+
+namespace CONTEXTMENU
+{
+
+  bool CEjectDisk::IsVisible(const CFileItem& item) const
+  {
+    return item.IsRemovable() && (item.IsDVD() || item.IsCDDA());
```
Comment (ksooo) [inline xbmc/ContextMenus.cpp]: Imo here a 
<pre>
#ifdef HAS_DVD_DRIVE
  return item.IsRemovable() && (item.IsDVD() || item.IsCDDA());
#else
  return false;
#endif
</pre>

would be a good choice to 100% prevent this item to appear on builds compiled without dvd drive support.

File: xbmc/ContextMenus.cpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "ContextMenus.h"
+#include "guilib/GUIWindowManager.h"
```
Comment (ksooo) [inline xbmc/ContextMenus.cpp]: Is this include really needed?

But I'm missing <code>#include "FileItem.h"</code>

File: xbmc/dialogs/GUIDialogContextMenu.cpp [hunk 3]
```diff
@@ -224,20 +222,6 @@ bool CGUIDialogContextMenu::SourcesMenu(const std::string &strType, const CFileI
 
 void CGUIDialogContextMenu::GetContextButtons(const std::string &type, const CFileItemPtr& item, CContextButtons &buttons)
 {
-  // Add buttons to the ContextMenu that should be visible for both sources and autosourced items
-  if (item && item->IsRemovable())
-  {
-    if (item->IsDVD() || item->IsCDDA())
-    {
-      buttons.Add(CONTEXT_BUTTON_EJECT_DISC, 13391);  // Eject/Load CD/DVD!
```
Comment (ksooo) [inline xbmc/dialogs/GUIDialogContextMenu.cpp]: CONTEXT_BUTTON_EJECT_DISC is no longer used now and schould be removed from dialogs/GUIDialogContextMenu.h

File: xbmc/dialogs/GUIDialogContextMenu.cpp [hunk 4]
```diff
@@ -224,20 +222,6 @@ bool CGUIDialogContextMenu::SourcesMenu(const std::string &strType, const CFileI
...
-    }
-    else // Must be HDD
-    {
-      buttons.Add(CONTEXT_BUTTON_EJECT_DRIVE, 13420);  // Eject Removable HDD!
```
Comment (ksooo) [inline xbmc/dialogs/GUIDialogContextMenu.cpp]: CONTEXT_BUTTON_EJECT_DRIVE is no longer used now and schould be removed from dialogs/GUIDialogContextMenu.h

**Review by ksooo — COMMENTED** [review]
Could you please address my comments and open a follow-up PR? Thnaks.

**Comment by Razzeee** [comment]
Nobody had anything to say for 1 and a half month. I don't exactly call this hurry.

Master is open for pre alpha again and I feel like having PRs to be perfect and not giving people the doubt to deliver follow up PRs is also hurting us more then helping. Just my 2 cent :)

File: xbmc/dialogs/GUIDialogContextMenu.cpp (hunk 3 above)
Comment (Razzeee) [inline xbmc/dialogs/GUIDialogContextMenu.cpp]: Good catch

File: xbmc/dialogs/GUIDialogContextMenu.cpp (hunk 4 above)
Comment (Razzeee) [inline xbmc/dialogs/GUIDialogContextMenu.cpp]: Good catch

**Comment by MartijnKaijser** [comment]
@Razzeee http://trac.kodi.tv/ticket/17143

**Comment by ksooo** [comment]
@Razzeee just discovered that this PR introduces another regression: "Eject" is no longer available in the context menu in Kodi's file manager. Before this, PR, for example, "Eject" appeared in the context menu of a mounted USB stick. Reason is that file manager does not use the new context menu system, but CGUIDialogContextMenu, where this PR removes the items in question from. :-/

**Comment by Razzeee** [comment]
So it seems that if "fails" when asking for the `m_listProvider` as it's null for that case.
Here: https://github.com/xbmc/xbmc/blob/c37e3c276434b5e8fb93c240cbc26e9e16502352/xbmc/guilib/GUIBaseContainer.cpp#L794

@tamland might have the insight on how to correctly tackle this

**Pull request merged**